File: src/test/scala/com/rbmhtechnology/example/dbreplica/cdc/AssetCdcInbound.scala
/*
 * Copyright (C) 2015 - 2016 Red Bull Media House GmbH <http://www.redbullmediahouse.com> - all rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rbmhtechnology.example.dbreplica.cdc

import com.rbmhtechnology.eventuate._
import com.rbmhtechnology.eventuate.ConcurrentVersions
import com.rbmhtechnology.eventuate.log.EventLogClock
import com.rbmhtechnology.example.dbreplica.domain._
import com.rbmhtechnology.example.dbreplica.event._
import com.rbmhtechnology.example.dbreplica.repository._
import com.rbmhtechnology.example.dbreplica.service._

import org.springframework.beans.factory.annotation._
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AssetCdcInbound @Autowired() (
  assetService: AssetService,
  assetEventRepository: AssetEventRepository,
  assetClockRepository: AssetClockRepository,
  progressRepository: ProgressRepository,
  assetCdcSettings: AssetCdcSettings) {

  import assetCdcSettings._

  @Transactional(readOnly = true)
  def readClock(clock: EventLogClock): EventLogClock =
    assetEventRepository.readClock(clock)

  @Transactional(readOnly = true)
  def readReplicationProgressAndVersion(logId: String, clock: EventLogClock): (Long, VectorTime) =
    (progressRepository.readReplicationProgress(logId), assetEventRepository.readClock(clock).versionVector)

  @Transactional
  def writeReplicationProgress(logId: String, progress: Long): Unit =
    progressRepository.writeReplicationProgress(logId, progress)

  @Transactional
  def handle(durableEvent: DurableEvent): Unit = {
    // ------------------------------------------
    //  TODO: implement batch replication writes
    // ------------------------------------------
    val sequenceNr = assetEventRepository.updateSequenceNr()
    val updatedEvent = durableEvent.copy(localLogId = logId, localSequenceNr = sequenceNr)

    updatedEvent.payload match {
      case e: AssetCreated => handleCreated(updatedEvent, e)
      case e: AssetEvent   => handleUpdated(updatedEvent, e)
    }
  }

  def handleCreated(durableEvent: DurableEvent, assetEvent: AssetCreated): Unit = {
    assetEventRepository.insert(assetEvent.assetId, durableEvent)
    assetClockRepository.insert(assetEvent.assetId, durableEvent.vectorTimestamp)
    assetService.create(eventProjection(null, assetEvent))
  }

  def handleUpdated(durableEvent: DurableEvent, assetEvent: AssetEvent): Unit = {
    val current = for {
      a <- assetService.find(assetEvent.assetId)
      c <- assetClockRepository.find(assetEvent.assetId)
    } yield (a, c)

    val (asset, clock) = current.get
    val (updatedAsset, updateTimestamp) =
      if (clock < durableEvent.vectorTimestamp) {
        (eventProjection(asset, assetEvent), durableEvent.vectorTimestamp)
      } else { // concurrent update
        val recoveredVersions = recoverVersions(assetEvent.assetId)
        val concurrentVersions = recoveredVersions.update(assetEvent, durableEvent.vectorTimestamp, durableEvent.systemTimestamp, durableEvent.processId)
        val conflictingVersions = concurrentVersions.all
        val resolvedVersion = resolveVersions(concurrentVersions).all.head
        assetService.notifySelected(resolvedVersion.value, conflictingVersions.map(_.value))
        (resolvedVersion.value, resolvedVersion.vectorTimestamp)
      }

    assetEventRepository.insert(assetEvent.assetId, durableEvent)
    assetClockRepository.update(assetEvent.assetId, clock.merge(updateTimestamp))
    assetService.update(updatedAsset)
  }

  private def recoverVersions(assetId: String): ConcurrentVersions[Asset, AssetEvent] = {
    assetEventRepository.findFor(assetId).foldLeft(zeroConcurrentVersions) {
      case (acc, upd) =>
        val updated = acc.update(upd.payload.asInstanceOf[AssetEvent], upd.vectorTimestamp, upd.systemTimestamp, upd.processId)
        if (updated.conflict) resolveVersions(updated) else updated
    }
  }

  private def resolveVersions(concurrentVersions: ConcurrentVersions[Asset, AssetEvent]): ConcurrentVersions[Asset, AssetEvent] = {
    val conflictingVersions = concurrentVersions.all.sortWith { (v1, v2) =>
      // Let version with higher timestamp win, in case of equal timestamps that with higher processId (= creator) wins
      if (v1.systemTimestamp == v2.systemTimestamp) v1.creator < v2.creator else v1.systemTimestamp < v2.systemTimestamp
    }
    concurrentVersions.resolve(conflictingVersions.last.vectorTimestamp)
  }

  private def zeroConcurrentVersions: ConcurrentVersions[Asset, AssetEvent] =
    ConcurrentVersions[Asset, AssetEvent](null, eventProjection)
}

File: src/test/java/com/rbmhtechnology/example/japi/dbreplica/cdc/AssetCdcInbound.java
/*
 * Copyright (C) 2015 - 2016 Red Bull Media House GmbH <http://www.redbullmediahouse.com> - all rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rbmhtechnology.example.japi.dbreplica.cdc;

import com.rbmhtechnology.eventuate.*;
import com.rbmhtechnology.eventuate.log.EventLogClock;
import com.rbmhtechnology.example.japi.dbreplica.domain.Asset;
import com.rbmhtechnology.example.japi.dbreplica.domain.AssetDoesNotExistException;
import com.rbmhtechnology.example.japi.dbreplica.event.AssetEvent;
import com.rbmhtechnology.example.japi.dbreplica.event.AssetEvent.AssetCreated;
import com.rbmhtechnology.example.japi.dbreplica.repository.AssetClockRepository;
import com.rbmhtechnology.example.japi.dbreplica.repository.AssetEventRepository;
import com.rbmhtechnology.example.japi.dbreplica.repository.ProgressRepository;
import com.rbmhtechnology.example.japi.dbreplica.service.AssetService;
import com.rbmhtechnology.example.japi.dbreplica.util.ScalaObjects;
import javaslang.Tuple;
import javaslang.Tuple2;
import javaslang.collection.Stream;
import javaslang.control.Match;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static com.rbmhtechnology.example.japi.dbreplica.cdc.AssetCdc.eventProjection;

@Service
public class AssetCdcInbound {
    @Autowired
    private AssetService assetService;
    @Autowired
    private AssetEventRepository assetEventRepository;
    @Autowired
    private AssetClockRepository assetClockRepository;
    @Autowired
    private ProgressRepository progressRepository;
    @Autowired
    private AssetCdcSettings assetCdcSettings;

    @Transactional(readOnly = true)
    public EventLogClock readClock(final EventLogClock clock) {
        return assetEventRepository.readClock(clock);
    }

    @Transactional(readOnly = true)
    public Tuple2<Long, VectorTime> readReplicationProgressAndVersion(final String logId, final EventLogClock clock) {
        return Tuple.of(progressRepository.readReplicationProgress(logId), readClock(clock).versionVector());
    }

    @Transactional
    public void writeReplicationProgress(final String logId, final Long progress) {
        progressRepository.writeReplicationProgress(logId, progress);
    }

    @Transactional
    public void handle(final DurableEvent durableEvent) {
        final Long sequenceNr = assetEventRepository.updateSequenceNr();
        final DurableEvent updatedEvent = ScalaObjects.DurableEvent.from(durableEvent).setLocalLogIdAndSequenceNr(getLogId(), sequenceNr);

        Match.of(updatedEvent.payload())
                .whenType(AssetCreated.class).thenRun(ev -> handleCreated(updatedEvent, ev))
                .whenType(AssetEvent.class).thenRun(ev -> handleUpdated(updatedEvent, ev));
    }

    private void handleCreated(final DurableEvent durableEvent, final AssetCreated assetEvent) {
        assetEventRepository.insert(assetEvent.assetId, durableEvent);
        assetClockRepository.insert(assetEvent.assetId, durableEvent.vectorTimestamp());
        assetService.create(eventProjection.apply(null, assetEvent));
    }

    private void handleUpdated(final DurableEvent durableEvent, final AssetEvent assetEvent) {
        final Tuple2<Asset, VectorTime> assetWithClock =
                assetService.find(assetEvent.assetId)
                        .flatMap(a -> assetClockRepository.find(assetEvent.assetId).map(c -> Tuple.of(a, c)))
                        .getOrElseThrow(() -> new AssetDoesNotExistException(assetEvent.assetId));

        final Asset asset = assetWithClock._1;
        final ScalaObjects.VectorTime clock = ScalaObjects.VectorTime.from(assetWithClock._2);

        if (clock.lt(durableEvent.vectorTimestamp())) {
            final Asset updatedAsset = eventProjection.apply(asset, assetEvent);
            persistUpdate(assetEvent, durableEvent, clock.get(), updatedAsset, durableEvent.vectorTimestamp());
        } else {
            final ConcurrentVersions<Asset, AssetEvent> recoveredVersions = recoverVersions(assetEvent.assetId);
            final ConcurrentVersions<Asset, AssetEvent> concurrentVersions = recoveredVersions.update(assetEvent, durableEvent.vectorTimestamp(), durableEvent.systemTimestamp(), durableEvent.processId());
            final List<Versioned<Asset>> conflictingVersions = concurrentVersions.getAll();
            final Versioned<Asset> resolvedVersion = Stream.ofAll(resolveVersions(concurrentVersions).getAll()).last();

            assetService.notifySelected(resolvedVersion.value(), Stream.ofAll(conflictingVersions).map(Versioned::value).toJavaList());
            persistUpdate(assetEvent, durableEvent, clock.get(), resolvedVersion.value(), resolvedVersion.vectorTimestamp());
        }
    }

    private void persistUpdate(final AssetEvent assetEvent, final DurableEvent durableEvent, final VectorTime clock, final Asset updatedAsset,
                               final VectorTime updateTimestamp) {
        assetEventRepository.insert(assetEvent.assetId, durableEvent);
        assetClockRepository.update(assetEvent.assetId, clock.merge(updateTimestamp));
        assetService.update(updatedAsset);
    }

    private ConcurrentVersions<Asset, AssetEvent> recoverVersions(final String assetId) {
        return Stream.ofAll(assetEventRepository.findFor(assetId))
                .foldLeft(
                        zeroConcurrentVersions(),
                        (acc, upd) -> {
                            final ConcurrentVersions<Asset, AssetEvent> updated = acc.update((AssetEvent) upd.payload(), upd.vectorTimestamp(), upd.systemTimestamp(), upd.processId());
                            if (updated.conflict()) {
                                return resolveVersions(updated);
                            } else {
                                return updated;
                            }
                        }
                );
    }

    private ConcurrentVersions<Asset, AssetEvent> resolveVersions(final ConcurrentVersions<Asset, AssetEvent> concurrentVersions) {
        final Versioned<Asset> winningVersion = Stream.ofAll(concurrentVersions.getAll())
                .sort((v1, v2) -> {
                    // Let version with higher timestamp win, in case of equal timestamps that with higher processId (= creator) wins
                    if (v1.systemTimestamp() == v2.systemTimestamp()) {
                        return v1.creator().compareTo(v2.creator());
                    } else {
                        return new Long(v1.systemTimestamp()).compareTo(v2.systemTimestamp());
                    }
                }).last();

        return concurrentVersions.resolve(winningVersion.vectorTimestamp());
    }

    private String getLogId() {
        return assetCdcSettings.getLogId();
    }

    private ConcurrentVersions<Asset, AssetEvent> zeroConcurrentVersions() {
        return ConcurrentVersionsTree.create(eventProjection);
    }
}
